File: backend/game_fetcher.py
"""
Game Fetcher for Personal Review System
Fetches games from Chess.com and Lichess APIs
"""

import aiohttp
import asyncio
from typing import List, Dict, Optional
from datetime import datetime, timedelta
import json
import os
import chess.pgn
from io import StringIO


class GameFetcher:
    """Fetches games from Chess.com and Lichess"""
    
    def __init__(self):
        self.cache_dir = "backend/cache/player_games"
        os.makedirs(self.cache_dir, exist_ok=True)
        
    async def fetch_games(
        self, 
        username: str, 
        platform: str = "chess.com",
        max_games: int = 100,
        months_back: int = 6
    ) -> List[Dict]:
        """
        Fetch games from specified platform
        
        Args:
            username: Player username
            platform: "chess.com", "lichess", or "combined"
            max_games: Maximum number of games to fetch
            months_back: How many months back to fetch
            
        Returns:
            List of game dictionaries with metadata and PGN
        """
        if platform == "combined":
            games_chess_com = await self._fetch_chess_com(username, max_games // 2, months_back)
            games_lichess = await self._fetch_lichess(username, max_games // 2, months_back)
            return games_chess_com + games_lichess
        elif platform == "chess.com":
            return await self._fetch_chess_com(username, max_games, months_back)
        elif platform == "lichess":
            return await self._fetch_lichess(username, max_games, months_back)
        else:
            raise ValueError(f"Unknown platform: {platform}")
    
    async def _fetch_chess_com(
        self, 
        username: str, 
        max_games: int,
        months_back: int
    ) -> List[Dict]:
        """Fetch games from Chess.com API"""
        games = []
        
        async with aiohttp.ClientSession() as session:
            # Get archives list
            archives_url = f"https://api.chess.com/pub/player/{username}/games/archives"
            
            try:
                async with session.get(archives_url) as response:
                    if response.status != 200:
                        print(f"Chess.com API error: {response.status}")
                        return []
                    
                    data = await response.json()
                    archives = data.get("archives", [])
                    
                    # Only fetch recent months
                    recent_archives = archives[-months_back:] if len(archives) > months_back else archives
                    
                    # Fetch games from each archive
                    for archive_url in reversed(recent_archives):  # Most recent first
                        if len(games) >= max_games:
                            break
                            
                        async with session.get(archive_url) as archive_response:
                            if archive_response.status == 200:
                                archive_data = await archive_response.json()
                                month_games = archive_data.get("games", [])
                                
                                for game_data in reversed(month_games):  # Most recent in month first
                                    if len(games) >= max_games:
                                        break
                                    
                                    game = self._parse_chess_com_game(game_data, username)
                                    if game:
                                        games.append(game)
                        
                        # Rate limiting
                        await asyncio.sleep(0.1)
            
            except Exception as e:
                print(f"Error fetching Chess.com games: {e}")
        
        return games
    
    def _parse_chess_com_game(self, game_data: Dict, username: str) -> Optional[Dict]:
        """Parse Chess.com game data into standard format"""
        try:
            pgn_text = game_data.get("pgn", "")
            if not pgn_text:
                return None
            
            # Parse PGN to extract metadata
            pgn_io = chess.pgn.read_game(StringIO(pgn_text))
            if not pgn_io:
                return None
            
            headers = pgn_io.headers
            
            # Determine player color and ratings
            white_player = headers.get("White", "").lower()
            black_player = headers.get("Black", "").lower()
            username_lower = username.lower()
            
            if username_lower in white_player:
                player_color = "white"
                player_rating = int(headers.get("WhiteElo", "0"))
                opponent_rating = int(headers.get("BlackElo", "0"))
                opponent_name = headers.get("Black", "Unknown")
            elif username_lower in black_player:
                player_color = "black"
                player_rating = int(headers.get("BlackElo", "0"))
                opponent_rating = int(headers.get("WhiteElo", "0"))
                opponent_name = headers.get("White", "Unknown")
            else:
                return None
            
            # Determine result from player perspective
            result_raw = headers.get("Result", "*")
            if result_raw == "1-0":
                result = "win" if player_color == "white" else "loss"
            elif result_raw == "0-1":
                result = "win" if player_color == "black" else "loss"
            elif result_raw == "1/2-1/2":
                result = "draw"
            else:
                result = "unknown"
            
            # Extract time control - convert to integer for consistent processing
            time_control_raw = game_data.get("time_control", headers.get("TimeControl", ""))
            try:
                # TimeControl in PGN is usually just seconds (e.g., "180") or format like "180+2"
                if isinstance(time_control_raw, str):
                    # Handle formats like "180", "180+2", or "600+5"
                    base_time = time_control_raw.split('+')[0] if '+' in time_control_raw else time_control_raw
                    time_control = int(base_time) if base_time.isdigit() else time_control_raw
                else:
                    time_control = time_control_raw
            except:
                time_control = time_control_raw
            
            time_class = game_data.get("time_class", "")
            
            # Categorize time control
            if time_class:
                if "bullet" in time_class:
                    time_category = "bullet"
                elif "blitz" in time_class:
                    time_category = "blitz"
                elif "rapid" in time_class:
                    time_category = "rapid"
                elif "daily" in time_class or "correspondence" in time_class:
                    time_category = "daily"
                else:
                    time_category = "classical"
            else:
                time_category = "unknown"
            
            # Accuracy data (Chess.com provides per-side accuracies for many games)
            accuracies = game_data.get("accuracies") or {}
            player_accuracy = None
            opponent_accuracy = None
            if isinstance(accuracies, dict):
                if player_color == "white":
                    player_accuracy = accuracies.get("white")
                    opponent_accuracy = accuracies.get("black")
                else:
                    player_accuracy = accuracies.get("black")
                    opponent_accuracy = accuracies.get("white")
            
            return {
                "game_id": game_data.get("url", "").split("/")[-1],
                "platform": "chess.com",
                "url": game_data.get("url", ""),
                "date": headers.get("Date", "").replace(".", "-"),
                "player_color": player_color,
                "player_rating": player_rating,
                "opponent_rating": opponent_rating,
                "opponent_name": opponent_name,
                "result": result,
                "opening": headers.get("ECOUrl", "").split("/")[-1] if headers.get("ECOUrl") else "",
                "eco": headers.get("ECO", ""),
                "termination": headers.get("Termination", ""),
                "time_control": time_control,
                "time_category": time_category,
                "pgn": pgn_text,
                "has_clock": "[%clk" in pgn_text,
                "accuracies": accuracies,
                "player_accuracy": player_accuracy,
                "opponent_accuracy": opponent_accuracy,
            }
        
        except Exception as e:
            print(f"Error parsing Chess.com game: {e}")
            return None
    
    async def _fetch_lichess(
        self, 
        username: str, 
        max_games: int,
        months_back: int
    ) -> List[Dict]:
        """Fetch games from Lichess API"""
        games = []
        
        # Calculate since timestamp (months_back)
        since_date = datetime.now() - timedelta(days=months_back * 30)
        since_ms = int(since_date.timestamp() * 1000)
        
        async with aiohttp.ClientSession() as session:
            url = f"https://lichess.org/api/games/user/{username}"
            params = {
                "max": max_games,
                "since": since_ms,
                "pgnInJson": "true",
                "clocks": "true",
                "evals": "false",
                "opening": "true"
            }
            
            headers = {
                "Accept": "application/x-ndjson"
            }
            
            try:
                async with session.get(url, params=params, headers=headers) as response:
                    if response.status != 200:
                        print(f"Lichess API error: {response.status}")
                        return []
                    
                    # Lichess returns NDJSON (newline-delimited JSON)
                    text = await response.text()
                    lines = text.strip().split("\n")
                    
                    for line in lines:
                        if not line:
                            continue
                        
                        try:
                            game_data = json.loads(line)
                            game = self._parse_lichess_game(game_data, username)
                            if game:
                                games.append(game)
                        except json.JSONDecodeError:
                            continue
            
            except Exception as e:
                print(f"Error fetching Lichess games: {e}")
        
        return games
    
    def _parse_lichess_game(self, game_data: Dict, username: str) -> Optional[Dict]:
        """Parse Lichess game data into standard format"""
        try:
            pgn_text = game_data.get("pgn", "")
            if not pgn_text:
                return None
            
            # Parse PGN
            pgn_io = chess.pgn.read_game(StringIO(pgn_text))
            if not pgn_io:
                return None
            
            headers = pgn_io.headers
            players = game_data.get("players", {})
            
            # Determine player color
            white_player = players.get("white", {}).get("user", {}).get("name", "").lower()
            black_player = players.get("black", {}).get("user", {}).get("name", "").lower()
            username_lower = username.lower()
            
            if username_lower == white_player:
                player_color = "white"
                player_rating = players.get("white", {}).get("rating", 0)
                opponent_rating = players.get("black", {}).get("rating", 0)
                opponent_name = players.get("black", {}).get("user", {}).get("name", "Unknown")
            elif username_lower == black_player:
                player_color = "black"
                player_rating = players.get("black", {}).get("rating", 0)
                opponent_rating = players.get("white", {}).get("rating", 0)
                opponent_name = players.get("white", {}).get("user", {}).get("name", "Unknown")
            else:
                return None
            
            # Determine result
            winner = game_data.get("winner", "")
            if winner == player_color:
                result = "win"
            elif winner == "":
                result = "draw"
            else:
                result = "loss"
            
            # Time control
            speed = game_data.get("speed", "")
            time_category = speed if speed in ["bullet", "blitz", "rapid", "classical", "correspondence"] else "unknown"
            
            # Opening
            opening_data = game_data.get("opening", {})
            opening_name = opening_data.get("name", "")
            eco = opening_data.get("eco", "")
            
            # Date
            created_at = game_data.get("createdAt", 0)
            date_str = datetime.fromtimestamp(created_at / 1000).strftime("%Y-%m-%d") if created_at else ""
            
            return {
                "game_id": game_data.get("id", ""),
                "platform": "lichess",
                "url": f"https://lichess.org/{game_data.get('id', '')}",
                "date": date_str,
                "player_color": player_color,
                "player_rating": player_rating,
                "opponent_rating": opponent_rating,
                "opponent_name": opponent_name,
                "result": result,
                "opening": opening_name,
                "eco": eco,
                "termination": game_data.get("status", ""),
                "time_control": f"{game_data.get('clock', {}).get('initial', 0)//60}+{game_data.get('clock', {}).get('increment', 0)}",
                "time_category": time_category,
                "pgn": pgn_text,
                "has_clock": "[%clk" in pgn_text
            }
        
        except Exception as e:
            print(f"Error parsing Lichess game: {e}")
            return None
    
    def cache_games(self, username: str, platform: str, games: List[Dict]) -> None:
        """Cache fetched games to file"""
        try:
            cache_file = os.path.join(self.cache_dir, f"{username}_{platform}.jsonl")
            with open(cache_file, "w") as f:
                for game in games:
                    f.write(json.dumps(game) + "\n")
            print(f"✓ Cached {len(games)} games for {username} ({platform})")
        except Exception as e:
            print(f"Warning: Could not cache games: {e}")
    
    def load_cached_games(self, username: str, platform: str) -> Optional[List[Dict]]:
        """Load games from cache if available"""
        try:
            cache_file = os.path.join(self.cache_dir, f"{username}_{platform}.jsonl")
            if not os.path.exists(cache_file):
                return None
            
            # Check if cache is recent (< 24 hours old)
            cache_age = datetime.now().timestamp() - os.path.getmtime(cache_file)
            if cache_age > 86400:  # 24 hours
                return None
            
            games = []
            with open(cache_file, "r") as f:
                for line in f:
                    if line.strip():
                        games.append(json.loads(line))
            
            print(f"✓ Loaded {len(games)} games from cache for {username} ({platform})")
            return games
        
        except Exception as e:
            print(f"Warning: Could not load cached games: {e}")
            return None

File: frontend/lib/gameFetcher.ts
/**
 * Frontend Game Fetcher
 * Fetches games directly from Chess.com and Lichess APIs
 */

import { Chess } from "chess.js";
import type { GameMetadata, GameFetchOptions } from "./gameReviewTypes";

/**
 * Fetch games from Chess.com API
 */
async function fetchChessComGames(
  username: string,
  options: GameFetchOptions
): Promise<GameMetadata[]> {
  const games: GameMetadata[] = [];
  const maxGames = options.max_games || 100;
  const monthsBack = options.months_back || 6;

  try {
    // Get archives list
    const archivesUrl = `https://api.chess.com/pub/player/${username}/games/archives`;
    const archivesResponse = await fetch(archivesUrl);

    if (!archivesResponse.ok) {
      throw new Error(`Chess.com API error: ${archivesResponse.status}`);
    }

    const archivesData = await archivesResponse.json();
    const archives = archivesData.archives || [];

    // Only fetch recent months
    const recentArchives =
      archives.length > monthsBack
        ? archives.slice(-monthsBack)
        : archives;

    // Fetch games from each archive (most recent first)
    for (const archiveUrl of recentArchives.reverse()) {
      if (games.length >= maxGames) break;

      // Rate limiting
      await new Promise((resolve) => setTimeout(resolve, 100));

      const archiveResponse = await fetch(archiveUrl);
      if (!archiveResponse.ok) continue;

      const archiveData = await archiveResponse.json();
      const monthGames = archiveData.games || [];

      for (const gameData of monthGames.reverse()) {
        if (games.length >= maxGames) break;

        const game = parseChessComGame(gameData, username);
        if (game) {
          // Apply filters
          if (matchesFilters(game, options)) {
            games.push(game);
          }
        }
      }
    }
  } catch (error) {
    console.error("Error fetching Chess.com games:", error);
    throw error;
  }

  return games;
}

/**
 * Parse Chess.com game data into standard format
 */
function parseChessComGame(
  gameData: any,
  username: string
): GameMetadata | null {
  try {
    const pgnText = gameData.pgn || "";
    if (!pgnText) return null;

    // Parse PGN to extract metadata
    const chess = new Chess();
    const pgn = chess.pgn({ pgn: pgnText });
    if (!pgn) return null;

    // Parse headers from PGN
    const headers: Record<string, string> = {};
    const headerLines = pgnText.split("\n").filter((line: string) =>
      line.startsWith("[")
    );
    for (const line of headerLines) {
      const match = line.match(/\[(\w+)\s+"([^"]+)"\]/);
      if (match) {
        headers[match[1]] = match[2];
      }
    }

    // Determine player color and ratings
    const whitePlayer = (headers.White || "").toLowerCase();
    const blackPlayer = (headers.Black || "").toLowerCase();
    const usernameLower = username.toLowerCase();

    let playerColor: "white" | "black";
    let playerRating: number;
    let opponentRating: number;
    let opponentName: string;

    if (usernameLower === whitePlayer) {
      playerColor = "white";
      playerRating = parseInt(headers.WhiteElo || "0", 10);
      opponentRating = parseInt(headers.BlackElo || "0", 10);
      opponentName = headers.Black || "Unknown";
    } else if (usernameLower === blackPlayer) {
      playerColor = "black";
      playerRating = parseInt(headers.BlackElo || "0", 10);
      opponentRating = parseInt(headers.WhiteElo || "0", 10);
      opponentName = headers.White || "Unknown";
    } else {
      return null;
    }

    // Determine result from player perspective
    const resultRaw = headers.Result || "*";
    let result: "win" | "loss" | "draw" | "unknown";
    if (resultRaw === "1-0") {
      result = playerColor === "white" ? "win" : "loss";
    } else if (resultRaw === "0-1") {
      result = playerColor === "black" ? "win" : "loss";
    } else if (resultRaw === "1/2-1/2") {
      result = "draw";
    } else {
      result = "unknown";
    }

    // Extract time control
    const timeControlRaw =
      gameData.time_control || headers.TimeControl || "";
    let timeControl: string | number = timeControlRaw;
    try {
      if (typeof timeControlRaw === "string") {
        const baseTime = timeControlRaw.split("+")[0];
        if (baseTime && /^\d+$/.test(baseTime)) {
          timeControl = parseInt(baseTime, 10);
        }
      }
    } catch {
      // Keep original value
    }

    const timeClass = gameData.time_class || "";
    let timeCategory = "unknown";
    if (timeClass.includes("bullet")) timeCategory = "bullet";
    else if (timeClass.includes("blitz")) timeCategory = "blitz";
    else if (timeClass.includes("rapid")) timeCategory = "rapid";
    else if (timeClass.includes("daily") || timeClass.includes("correspondence"))
      timeCategory = "daily";
    else if (timeClass) timeCategory = "classical";

    // Accuracy data
    const accuracies = gameData.accuracies || {};
    const playerAccuracy =
      playerColor === "white"
        ? accuracies.white
        : accuracies.black;
    const opponentAccuracy =
      playerColor === "white"
        ? accuracies.black
        : accuracies.white;

    return {
      game_id: (gameData.url || "").split("/").pop() || "",
      platform: "chess.com",
      url: gameData.url || "",
      date: (headers.Date || "").replace(/\./g, "-"),
      player_color: playerColor,
      player_rating: playerRating,
      opponent_rating: opponentRating,
      opponent_name: opponentName,
      result,
      opening: headers.ECOUrl
        ? headers.ECOUrl.split("/").pop() || ""
        : "",
      eco: headers.ECO || "",
      termination: headers.Termination || "",
      time_control: timeControl,
      time_category: timeCategory,
      pgn: pgnText,
      has_clock: pgnText.includes("[%clk"),
      accuracies,
      player_accuracy: playerAccuracy,
      opponent_accuracy: opponentAccuracy,
    };
  } catch (error) {
    console.error("Error parsing Chess.com game:", error);
    return null;
  }
}

/**
 * Fetch games from Lichess API
 */
async function fetchLichessGames(
  username: string,
  options: GameFetchOptions
): Promise<GameMetadata[]> {
  const games: GameMetadata[] = [];
  const maxGames = options.max_games || 100;
  const monthsBack = options.months_back || 6;

  try {
    // Calculate since timestamp
    const sinceDate = new Date();
    sinceDate.setMonth(sinceDate.getMonth() - monthsBack);
    const sinceMs = sinceDate.getTime();

    const url = `https://lichess.org/api/games/user/${username}`;
    const params = new URLSearchParams({
      max: String(maxGames),
      since: String(sinceMs),
      pgnInJson: "true",
      clocks: "true",
      evals: "false",
      opening: "true",
    });

    const response = await fetch(`${url}?${params}`, {
      headers: {
        Accept: "application/x-ndjson",
      },
    });

    if (!response.ok) {
      throw new Error(`Lichess API error: ${response.status}`);
    }

    // Lichess returns NDJSON (newline-delimited JSON)
    const text = await response.text();
    const lines = text.trim().split("\n");

    for (const line of lines) {
      if (!line.trim()) continue;

      try {
        const gameData = JSON.parse(line);
        const game = parseLichessGame(gameData, username);
        if (game && matchesFilters(game, options)) {
          games.push(game);
        }
      } catch (e) {
        console.warn("Failed to parse Lichess game line:", e);
      }
    }
  } catch (error) {
    console.error("Error fetching Lichess games:", error);
    throw error;
  }

  return games;
}

/**
 * Parse Lichess game data into standard format
 */
function parseLichessGame(
  gameData: any,
  username: string
): GameMetadata | null {
  try {
    const pgnText = gameData.pgn || "";
    if (!pgnText) return null;

    const players = gameData.players || {};
    const whitePlayer = (
      players.white?.user?.name || ""
    ).toLowerCase();
    const blackPlayer = (
      players.black?.user?.name || ""
    ).toLowerCase();
    const usernameLower = username.toLowerCase();

    let playerColor: "white" | "black";
    let playerRating: number;
    let opponentRating: number;
    let opponentName: string;

    if (usernameLower === whitePlayer) {
      playerColor = "white";
      playerRating = players.white?.rating || 0;
      opponentRating = players.black?.rating || 0;
      opponentName = players.black?.user?.name || "Unknown";
    } else if (usernameLower === blackPlayer) {
      playerColor = "black";
      playerRating = players.black?.rating || 0;
      opponentRating = players.white?.rating || 0;
      opponentName = players.white?.user?.name || "Unknown";
    } else {
      return null;
    }

    // Determine result
    const winner = gameData.winner || "";
    let result: "win" | "loss" | "draw" | "unknown";
    if (winner === playerColor) {
      result = "win";
    } else if (winner === "") {
      result = "draw";
    } else {
      result = "loss";
    }

    // Time control
    const speed = gameData.speed || "";
    const timeCategory =
      ["bullet", "blitz", "rapid", "classical", "correspondence"].includes(
        speed
      )
        ? speed
        : "unknown";

    const clock = gameData.clock || {};
    const timeControl = clock.initial
      ? `${Math.floor(clock.initial / 60)}+${clock.increment || 0}`
      : "";

    // Opening
    const openingData = gameData.opening || {};
    const openingName = openingData.name || "";
    const eco = openingData.eco || "";

    // Date
    const createdAt = gameData.createdAt || 0;
    const dateStr = createdAt
      ? new Date(createdAt).toISOString().split("T")[0]
      : "";

    return {
      game_id: gameData.id || "",
      platform: "lichess",
      url: `https://lichess.org/${gameData.id || ""}`,
      date: dateStr,
      player_color: playerColor,
      player_rating: playerRating,
      opponent_rating: opponentRating,
      opponent_name: opponentName,
      result,
      opening: openingName,
      eco,
      termination: gameData.status || "",
      time_control: timeControl,
      time_category: timeCategory,
      pgn: pgnText,
      has_clock: pgnText.includes("[%clk"),
    };
  } catch (error) {
    console.error("Error parsing Lichess game:", error);
    return null;
  }
}

/**
 * Check if game matches filters
 */
function matchesFilters(
  game: GameMetadata,
  options: GameFetchOptions
): boolean {
  // Opponent filter
  if (
    options.opponent &&
    !game.opponent_name.toLowerCase().includes(options.opponent.toLowerCase())
  ) {
    return false;
  }

  // Color filter
  if (options.color && game.player_color !== options.color) {
    return false;
  }

  // Result filter
  if (
    options.result_filter &&
    options.result_filter !== "all" &&
    game.result !== options.result_filter
  ) {
    return false;
  }

  // Time control filter
  if (
    options.time_control &&
    options.time_control !== "all" &&
    game.time_category !== options.time_control
  ) {
    return false;
  }

  // Opening ECO filter
  if (options.opening_eco && game.eco !== options.opening_eco) {
    return false;
  }

  // Rating filters
  if (
    options.min_opponent_rating &&
    game.opponent_rating < options.min_opponent_rating
  ) {
    return false;
  }
  if (
    options.max_opponent_rating &&
    game.opponent_rating > options.max_opponent_rating
  ) {
    return false;
  }

  // Date filters
  if (options.date_from && game.date < options.date_from) {
    return false;
  }
  if (options.date_to && game.date > options.date_to) {
    return false;
  }

  return true;
}

/**
 * Main function to fetch games
 */
export async function fetchGames(
  options: GameFetchOptions
): Promise<GameMetadata[]> {
  const { platform, username } = options;

  if (platform === "chess.com") {
    return fetchChessComGames(username, options);
  } else if (platform === "lichess") {
    return fetchLichessGames(username, options);
  } else {
    throw new Error(`Unknown platform: ${platform}`);
  }
}
